File: cpp_agent/cpp_board_lib/src/lib/board.cpp
#include <iostream>
#include <istream>
#include <vector>
#include <map>
#include <random>
#include <utility>
#include <lib/board.h>
#include <lib/utils.h>

Board::Board(int r, int c, int seed, int numLinesTaken) : rows(r), cols(c) {
    mt19937 gen(seed);

    horizontalLines.resize(rows, vector<int>(cols - 1, 0));
    verticalLines.resize(rows - 1, vector<int>(cols, 0));

    int numHorizontalLines = gen() % (numLinesTaken + 1);
    int numVerticalLines = numLinesTaken - numHorizontalLines;
    for (int i = 0; i < numHorizontalLines; ++i) {
        int row = gen() % rows;
        int col = gen() % (cols - 1);
        if (horizontalLines[row][col])
            i--;
        else
            horizontalLines[row][col] = 1;
    }
    for (int i = 0; i < numVerticalLines; ++i) {
        int row = gen() % (rows - 1);
        int col = gen() % cols;
        if (verticalLines[row][col])
            i--;
        else
            verticalLines[row][col] = 1;
    }

    gridOwner.resize(
        rows - 1,
        vector<GridOwner>(cols - 1,
                          GridOwner::UNSPECIFIED)); // 0 indicates no owner
    int numGridsTaken = 0;
    for (int i = 0; i < rows - 1; ++i) {
        for (int j = 0; j < cols - 1; ++j) {
            if (horizontalLines[i][j] == 1 && horizontalLines[i + 1][j] == 1 &&
                verticalLines[i][j] == 1 && verticalLines[i][j + 1] == 1) {
                gridOwner[i][j] = GridOwner::PRE_FILLED;
                numGridsTaken++;
            }
        }
    }

    numEmptyGrids = (rows - 1) * (cols - 1) - numGridsTaken;
    numHorizontalLinesLeft = (rows * (cols - 1)) - numHorizontalLines;
    numVerticalLinesLeft = ((rows - 1) * cols) - numVerticalLines;
    scores.clear();
}

Board::Board(int r, int c, vector<vector<int>> hLines,
             vector<vector<int>> vLines, vector<vector<GridOwner>> gOwner)
    : rows(r), cols(c), horizontalLines(hLines), verticalLines(vLines),
      gridOwner(gOwner) {
    cerr << "Initializing Board with given lines and owner." << endl;
    numEmptyGrids = 0;
    numHorizontalLinesLeft = 0;
    numVerticalLinesLeft = 0;
    scores.clear();
    for (int i = 0; i < rows - 1; ++i) {
        for (int j = 0; j < cols - 1; ++j) {
            cerr << "Grid (" << i << ", " << j
                 << ") owned by: " << gOwner[i][j] << endl;
            if (gridOwner[i][j] == GridOwner::UNSPECIFIED) {
                cerr << "Grid (" << i << ", " << j << ") is empty." << endl;
                numEmptyGrids++;
            }
        }
    }
    for (int i = 0; i < rows; ++i) {
        for (int j = 0; j < cols - 1; ++j) {
            if (horizontalLines[i][j] == 0) {
                numHorizontalLinesLeft++;
            }
        }
    }
    for (int i = 0; i < rows - 1; ++i) {
        for (int j = 0; j < cols; ++j) {
            if (verticalLines[i][j] == 0) {
                numVerticalLinesLeft++;
            }
        }
    }
}

bool Board::isValidMove(Move move) const {
    if (move.isHorizontal) {
        if (move.row < 0 || move.row >= rows || move.col < 0 ||
            move.col >= cols - 1)
            return false;
        return horizontalLines[move.row][move.col] ==
               0; // Check if the line is not already drawn
    } else {
        if (move.row < 0 || move.row >= rows - 1 || move.col < 0 ||
            move.col >= cols)
            return false;
        return verticalLines[move.row][move.col] ==
               0; // Check if the line is not already drawn
    }
}

bool Board::requiresContinuation(Move move) const {
    return !isCompletingMove(move) && isCapturingMove(move);
}

bool Board::isCompletingMove(Move move) const {
    return (numHorizontalLinesLeft + numVerticalLinesLeft) ==
           1; // Presumes move is valid
}

bool Board::isCapturingMove(Move move) const {
    return !getCapturingGrids(*this, move).empty();
}

bool Board::makeMove(Move move, PlayerSide moveSide) {
    bool isCapturing;
    bool isCompleting = isCompletingMove(move);

    // mark the grid as owned by the player if capturing
    for (auto [gridRow, gridCol] : getCapturingGrids(*this, move)) {
        gridOwner[gridRow][gridCol] = moveSide == PlayerSide::FIRST_PLAYER
                                          ? GridOwner::FIRST_PLAYER
                                          : GridOwner::SECOND_PLAYER;
        isCapturing = true;
        numEmptyGrids--;
        scores[moveSide]++;
    }

    // Mark the line as drawn

    if (move.isHorizontal) {
        horizontalLines[move.row][move.col] =
            1; // Mark the horizontal line as drawn
        numHorizontalLinesLeft--;
    } else {
        verticalLines[move.row][move.col] =
            1; // Mark the vertical line as drawn
        numVerticalLinesLeft--;
    }
    return isCapturing && !isCompleting;
}

bool Board::isCompleted() const { return numEmptyGrids == 0; }

const map<PlayerSide, int> &Board::getScores() const { return scores; }

vector<Move> Board::getValidMoves() const {
    vector<Move> validMoves;
    for (int r = 0; r < rows; ++r) {
        for (int c = 0; c < cols - 1; ++c) {
            if (horizontalLines[r][c] ==
                0) // If the horizontal line is not drawn
            {
                validMoves.emplace_back(r, c, true);
            }
        }
    }
    for (int r = 0; r < rows - 1; ++r) {
        for (int c = 0; c < cols; ++c) {
            if (verticalLines[r][c] == 0) // If the vertical line is not drawn
            {
                validMoves.emplace_back(r, c, false);
            }
        }
    }
    return validMoves;
}

ostream &operator<<(std::ostream &os, const Board &board) {
    os << board.rows << " " << board.cols << endl;
    for (const auto &row : board.horizontalLines) {
        for (int line : row) {
            os << line << " ";
        }
        os << endl;
    }
    for (const auto &row : board.verticalLines) {
        for (int line : row) {
            os << line << " ";
        }
        os << endl;
    }
    for (const auto &row : board.gridOwner) {
        for (GridOwner owner : row) {
            os << owner << " ";
        }
        os << endl;
    }
    return os;
}

istream &operator>>(std::istream &is, Board &board) {
    is >> board.rows >> board.cols;
    board.horizontalLines.resize(board.rows, vector<int>(board.cols - 1));
    board.verticalLines.resize(board.rows - 1, vector<int>(board.cols));
    board.gridOwner.resize(
        board.rows - 1,
        vector<GridOwner>(board.cols - 1, GridOwner::UNSPECIFIED));
    board.numEmptyGrids = 0;
    board.numHorizontalLinesLeft = 0;
    board.numVerticalLinesLeft = 0;
    board.scores.clear();
    for (int i = 0; i < board.rows; ++i) {
        for (int j = 0; j < board.cols - 1; ++j) {
            is >> board.horizontalLines[i][j];
            if (board.horizontalLines[i][j] == 0) {
                board.numHorizontalLinesLeft++;
            }
        }
    }
    for (int i = 0; i < board.rows - 1; ++i) {
        for (int j = 0; j < board.cols; ++j) {
            is >> board.verticalLines[i][j];
            if (board.verticalLines[i][j] == 0) {
                board.numVerticalLinesLeft++;
            }
        }
    }
    for (int i = 0; i < board.rows - 1; ++i) {
        for (int j = 0; j < board.cols - 1; ++j) {
            is >> board.gridOwner[i][j];
            switch (board.gridOwner[i][j]) {
            case GridOwner::UNSPECIFIED:
                board.numEmptyGrids++;
                break;
            case GridOwner::FIRST_PLAYER:
                board.scores[PlayerSide::FIRST_PLAYER]++;
                break;
            case GridOwner::SECOND_PLAYER:
                board.scores[PlayerSide::SECOND_PLAYER]++;
                break;
            case GridOwner::PRE_FILLED:
                // don nothing:
                break;
            }
        }
    }
    return is;
}
std::istream &operator>>(istream &is, PlayerSide &ps) {
    is >> (int &)ps;
    return is;
}

std::ostream &operator<<(ostream &os, const PlayerSide &ps) {
    os << (int)ps;
    return os;
}

std::istream &operator>>(istream &is, GridOwner &go) {
    is >> (int &)go;
    return is;
}

std::ostream &operator<<(ostream &os, const GridOwner &go) {
    os << (int)go;
    return os;
}

File: python_agent_4/board.py
from __future__ import annotations

from copy import deepcopy
from enum import IntEnum
from typing import Dict, Iterable, List, Sequence, Tuple

from .move import Move
from .token_stream import TokenStream


class PlayerSide(IntEnum):
    FIRST_PLAYER = 1
    SECOND_PLAYER = 2

    def opponent(self) -> "PlayerSide":
        return PlayerSide.SECOND_PLAYER if self is PlayerSide.FIRST_PLAYER else PlayerSide.FIRST_PLAYER


class GridOwner(IntEnum):
    UNSPECIFIED = 0
    FIRST_PLAYER = 1
    SECOND_PLAYER = 2
    PRE_FILLED = 3

    @staticmethod
    def from_side(side: PlayerSide) -> "GridOwner":
        if side is PlayerSide.FIRST_PLAYER:
            return GridOwner.FIRST_PLAYER
        if side is PlayerSide.SECOND_PLAYER:
            return GridOwner.SECOND_PLAYER
        raise ValueError(f"Unsupported player side: {side}")


class Board:
    """Game board state mirroring the behaviour of the C++ reference."""

    def __init__(
        self,
        rows: int,
        cols: int,
        horizontal_lines: Sequence[Sequence[int]],
        vertical_lines: Sequence[Sequence[int]],
        grid_owner: Sequence[Sequence[int | GridOwner]],
    ) -> None:
        self.rows = rows
        self.cols = cols
        self.horizontal_lines: List[List[int]] = [list(row) for row in horizontal_lines]
        self.vertical_lines: List[List[int]] = [list(row) for row in vertical_lines]
        self.grid_owner: List[List[GridOwner]] = [
            [cell if isinstance(cell, GridOwner) else GridOwner(cell) for cell in row]
            for row in grid_owner
        ]

        self.scores: Dict[PlayerSide, int] = {
            PlayerSide.FIRST_PLAYER: 0,
            PlayerSide.SECOND_PLAYER: 0,
        }
        self.num_empty_grids = 0
        self.num_horizontal_lines_left = 0
        self.num_vertical_lines_left = 0
        self._recompute_metadata()

    @classmethod
    def from_token_stream(cls, tokens: TokenStream) -> "Board":
        rows = tokens.next_int()
        cols = tokens.next_int()
        horizontal = [[tokens.next_int() for _ in range(cols - 1)] for _ in range(rows)]
        vertical = [[tokens.next_int() for _ in range(cols)] for _ in range(rows - 1)]
        owners = [[tokens.next_int() for _ in range(cols - 1)] for _ in range(rows - 1)]
        return cls(rows, cols, horizontal, vertical, owners)

    def clone(self) -> "Board":
        return Board(
            self.rows,
            self.cols,
            deepcopy(self.horizontal_lines),
            deepcopy(self.vertical_lines),
            deepcopy(self.grid_owner),
        )

    def _recompute_metadata(self) -> None:
        self.num_horizontal_lines_left = sum(1 for row in self.horizontal_lines for cell in row if cell == 0)
        self.num_vertical_lines_left = sum(1 for row in self.vertical_lines for cell in row if cell == 0)
        self.num_empty_grids = 0
        self.scores[PlayerSide.FIRST_PLAYER] = 0
        self.scores[PlayerSide.SECOND_PLAYER] = 0
        for row in self.grid_owner:
            for owner in row:
                if owner is GridOwner.UNSPECIFIED:
                    self.num_empty_grids += 1
                elif owner is GridOwner.FIRST_PLAYER:
                    self.scores[PlayerSide.FIRST_PLAYER] += 1
                elif owner is GridOwner.SECOND_PLAYER:
                    self.scores[PlayerSide.SECOND_PLAYER] += 1

    def is_valid_move(self, move: Move) -> bool:
        if move.is_horizontal:
            if move.row < 0 or move.row >= self.rows:
                return False
            if move.col < 0 or move.col >= self.cols - 1:
                return False
            return self.horizontal_lines[move.row][move.col] == 0
        if move.row < 0 or move.row >= self.rows - 1:
            return False
        if move.col < 0 or move.col >= self.cols:
            return False
        return self.vertical_lines[move.row][move.col] == 0

    def requires_continuation(self, move: Move) -> bool:
        return self.is_capturing_move(move) and not self.is_completing_move(move)

    def is_completing_move(self, move: Move) -> bool:
        return (self.num_horizontal_lines_left + self.num_vertical_lines_left) == 1

    def is_capturing_move(self, move: Move) -> bool:
        return bool(get_capturing_grids(self, move))

    def make_move(self, move: Move, side: PlayerSide) -> bool:
        if not self.is_valid_move(move):
            raise ValueError(f"Invalid move attempted: {move}")

        capturing_grids = get_capturing_grids(self, move)
        is_completing = self.is_completing_move(move)

        for grid_row, grid_col in capturing_grids:
            previous_owner = self.grid_owner[grid_row][grid_col]
            if previous_owner is GridOwner.UNSPECIFIED:
                self.num_empty_grids -= 1
            elif previous_owner in (GridOwner.FIRST_PLAYER, GridOwner.SECOND_PLAYER):
                previous_side = PlayerSide(previous_owner.value)
                self.scores[previous_side] -= 1
            owner = GridOwner.from_side(side)
            self.grid_owner[grid_row][grid_col] = owner
            self.scores[side] += 1

        if move.is_horizontal:
            self.horizontal_lines[move.row][move.col] = 1
            self.num_horizontal_lines_left -= 1
        else:
            self.vertical_lines[move.row][move.col] = 1
            self.num_vertical_lines_left -= 1

        return bool(capturing_grids) and not is_completing

    def is_completed(self) -> bool:
        return self.num_empty_grids == 0

    def get_scores(self) -> Dict[PlayerSide, int]:
        return dict(self.scores)

    def get_valid_moves(self) -> List[Move]:
        moves: List[Move] = []
        for r in range(self.rows):
            for c in range(self.cols - 1):
                if self.horizontal_lines[r][c] == 0:
                    moves.append(Move(r, c, True))
        for r in range(self.rows - 1):
            for c in range(self.cols):
                if self.vertical_lines[r][c] == 0:
                    moves.append(Move(r, c, False))
        return moves


def get_capturing_grids(board: Board, move: Move) -> List[Tuple[int, int]]:
    capturing: List[Tuple[int, int]] = []
    if _is_capturing_above(board, move):
        capturing.append((move.row - 1, move.col))
    if _is_capturing_below(board, move):
        capturing.append((move.row, move.col))
    if _is_capturing_left(board, move):
        capturing.append((move.row, move.col - 1))
    if _is_capturing_right(board, move):
        capturing.append((move.row, move.col))
    return capturing


def _line_drawn(value: int) -> bool:
    return value != 0


def _is_capturing_above(board: Board, move: Move) -> bool:
    if not move.is_horizontal or move.row <= 0:
        return False
    return (
        _line_drawn(board.vertical_lines[move.row - 1][move.col])
        and _line_drawn(board.vertical_lines[move.row - 1][move.col + 1])
        and _line_drawn(board.horizontal_lines[move.row - 1][move.col])
    )


def _is_capturing_below(board: Board, move: Move) -> bool:
    if not move.is_horizontal or move.row >= board.rows - 1:
        return False
    return (
        _line_drawn(board.vertical_lines[move.row][move.col])
        and _line_drawn(board.vertical_lines[move.row][move.col + 1])
        and _line_drawn(board.horizontal_lines[move.row + 1][move.col])
    )


def _is_capturing_left(board: Board, move: Move) -> bool:
    if move.is_horizontal or move.col <= 0:
        return False
    return (
        _line_drawn(board.horizontal_lines[move.row][move.col - 1])
        and _line_drawn(board.horizontal_lines[move.row + 1][move.col - 1])
        and _line_drawn(board.vertical_lines[move.row][move.col - 1])
    )


def _is_capturing_right(board: Board, move: Move) -> bool:
    if move.is_horizontal or move.col >= board.cols - 1:
        return False
    return (
        _line_drawn(board.horizontal_lines[move.row][move.col])
        and _line_drawn(board.horizontal_lines[move.row + 1][move.col])
        and _line_drawn(board.vertical_lines[move.row][move.col + 1])
    )
